genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobKillServiceImpl.java

```diff
@@ -110,4 +111,20 @@ public void killJob(final @NotBlank(message = "No job id entered. Unable to kill
             log.info("Agent notified for killing job {}", jobId);
         }
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onConnected(@NotNull final String jobId) {
+        //No op
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onDisconnected(@NotNull final String jobId) {
+        parkedJobKillResponseObservers.remove(jobId);
```

Comment:
Possible race condition here.
Agent 1 disconnects, reconnects to the same server.
Then the server processes the first onDisconnect.
The second connection (still valid) will be removed from the map.

No suggestions on how to address this... I haven't looked at details and how we could make this better.
But you should probably address this scenario.

Reply:
First thing that comes to my mind is that essentially we need some kind of ordering - a `disconnect` event should remove only the response observer which was created in the map before the `disconnect` event. So the map of response observers can also have a `creation timestamp`. That can be checked before removing. We should possibly whiteboard this.